### PR TITLE
Add interactive command to get a random problem:

### LINKED ISSUE #78: Add a command to get a random leetcode challenge

### PR DESCRIPTION
**Pull request by kajahno**
* Also refactor the way we get the data from Discord
* Added the static problem set from Leetcode (about 2.6k questions)
* Added a few ad-hoc commands, useful for testing

Closes #78

### CODE REVIEW

**Review by nramirez — COMMENTED** [review]
Only left a few comments/suggestions.

A quick question, do we really need to add this file?
```
api/leetcode/static/problemSets.json
```

My only concern is that our current pattern is making api/leetcode.js hard to follow, but we can probably address that in a different PR.

File: api/leetcode.js [hunk 1]
```diff
@@ -21,64 +28,218 @@ export class Leetcode {
...
+        Builds a discord Leetcode challenge simplified message
+    */
+    _buildSimplifiedProblemMessage = async (problem) => {
+        return new MessageEmbed()
+            .setColor(problem.color)
+            .setTitle(`${problem.questionFrontendId}. ${problem.title}`)
+            .setURL(`${URL}${problem.link}`)
+            .addFields({ name: "Remaining time", value: `${problem.remainingTimeMessage}`, inline: false })
+            .setFooter({ text: "Time to code 🔥👨‍💻🔥" });
+    }
```
[the remark this answers is not in the document]
Comment (kajahno) [reply api/leetcode.js:211]: I think I'll leave it here for now, I'm not sure how this is going to change in the future (we can change it when the time comes!)

File: api/leetcode.js [hunk 2]
```diff
@@ -21,64 +28,218 @@ export class Leetcode {
...
+}
+
+
+export class LeetcodeData {
+
+    constructor() {}
+
+    getProblemSetData = async () => {
+        // TODO: find a way to memoize this. It takes too long.
+        const { data: { problemsetQuestionList: { questions } } } = await getProblemSet();
```
[the remark this answers is not in the document]
Comment (kajahno) [reply api/leetcode.js:221]: Let me think about this and get back to you!

[the remark this answers is not in the document]
Comment (kajahno) [reply api/leetcode.js]: Done

File: api/leetcode-data.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
+import { dailyGetLeetcodeData, weeklyGetLeetcodeData, getProblemSet } from "./leetcode/index.js";
+import logger from "../logging.js";
+import staticProblemSets from './leetcode/static/problemSets.json' assert { type: "json" };
+
+export class LeetcodeData {
+
+    constructor() {}
```
Comment (nramirez) [inline api/leetcode-data.js:7]: nit: is there any reason you want to have this empty constructor? I usually don't have them around unless there's some important initialization.

File: api/leetcode-data.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    getProblemSetData = async (limit, skip) => {
+        const { data: { problemsetQuestionList: { questions } } } = await getProblemSet(limit, skip);
```
Comment (nramirez) [inline api/leetcode-data.js:10]: this setup is a little dangerous, especially because we don't use typescript; someone in the future could change the return type in `getProblemSet` without noticing that they're breaking some code that it is assuming this data structure.

A slightly better approach is to get the `data` in an object as a whole, and then have the null check for each one of the nested fields, it would be something like this:

```js
const resposne = await getProblemSet(limit, skip);

if (!response?.data?.problemsetQuestionList?.questions) {
    logger.error("Unable to fetch problemsetQuestionList");
    return;
}

```

This is technically the same thing, but exception free

File: api/leetcode-data.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        if (!questions) {
+            logger.error("Unable to fetch problemsetQuestionList");
+            return;
+        }
+        return questions
+    };
+
+    getNumProblems = async () => {
+        const { data: { problemsetQuestionList: { total } } } = await getProblemSet();
+        if (!total) {
```
Comment (nramirez) [inline api/leetcode-data.js:20]: ditto, but up to you is you want to live with this instead becase is clearer for you.

File: app.js [hunk 6]
```diff
@@ -97,6 +100,18 @@ const runDaemon = async () => {
...
+            // "random problem" guild command
+            if (name === "randomproblem") {
+                const message = await leetcode.getRandomProblemMessage();
+                return res.send({
+                    type: InteractionResponseType.CHANNEL_MESSAGE_WITH_SOURCE,
+                    data: message,
+                });
+            }
+
+            // TODO: "problem from set" guild command
```
Comment (nramirez) [inline app.js:113]: you could maybe link this to a ticket, TODO(#randomnumber): Add problem set guild command

**Review by nramirez — APPROVED** [review]
LGTM, some suggestions, but looks sane to me.

File: api/leetcode-data.js (hunk 4 above)
Comment (kajahno) [reply api/leetcode-data.js:10]: Done, thanks!

File: app.js (hunk 6 above)
Comment (kajahno) [reply app.js:113]: Done, created #80
